In [ ]:
!pip install -q pyomo

     |████████████████████████████████| 9.7 MB 5.5 MB/s 
     |████████████████████████████████| 49 kB 5.6 MB/s 


In [ ]:
from pyomo.environ import * 

In [ ]:
import numpy as np

ANSWER2:(a)Let Indian Air Force decides to purchage $x_1 ,x_2,x_3,x_4,x_5 $ number of drones of models Shrey,Dharan,Satva,Adhya,Pravat respectfully.Accornding to question:

  
Obj: max $ 11x_1+13x_2+9x_3+15x_4+12x_5 \\
 6.5x_1+8x_2+10.5x_3+11.75x_4+14.5x_5 \leq 78 \\
 5.75x_1+4.5x_2+3.75x_3+4.85x_4+3.45x_5 \leq 84 \\ $ 

 Here all variables   $x_1 ,x_2,x_3,x_4,x_5 $  must be integers since all denote the number of drones of any specific model.

In [ ]:
import pandas as pd

In [ ]:
data_csvfile = pd.read_csv('lab4_ex3.csv')

In [ ]:
print (type(data_csvfile))

<class 'pandas.core.frame.DataFrame'>


In [ ]:
data_csvfile.columns

Index(['Drone model', 'Carrying capacity (kgs)', 'Cost (in crores of rupees)',
       'Parking space (sq.m.)'],
      dtype='object')

In [ ]:
data_csvfile['Drone model']

0     Shrey
1    Dharan
2     Satva
3     Adhya
4    Pravat
Name: Drone model, dtype: object

In [ ]:
data_csvfile['Carrying capacity (kgs)']

0    11
1    13
2     9
3    15
4    12
Name: Carrying capacity (kgs), dtype: int64

In [ ]:
data_csvfile['Cost (in crores of rupees)']

0     6.50
1     8.00
2    10.50
3    11.75
4    14.50
Name: Cost (in crores of rupees), dtype: float64

In [ ]:
data_csvfile['Parking space (sq.m.)']

0    5.75
1    4.50
2    3.75
3    4.85
4    3.45
Name: Parking space (sq.m.), dtype: float64

In [ ]:
data_csvfile.index

RangeIndex(start=0, stop=5, step=1)

In [ ]:
len(data_csvfile.index)

5

In [ ]:
# create a model
model = ConcreteModel()

In [ ]:
N = len(data_csvfile.index)

In [ ]:
M = 5

In [ ]:
col_indices = range(N)

In [ ]:
model.x = Var(col_indices,domain=NonNegativeIntegers)

In [ ]:
model.objective = Objective(expr=summation(data_csvfile['Carrying capacity (kgs)'], model.x),sense=maximize)

In [ ]:
model.constraints = ConstraintList()

In [ ]:
model.constraints.add(summation(data_csvfile['Cost (in crores of rupees)'], model.x) <= 78)
model.constraints.add(summation(data_csvfile['Parking space (sq.m.)'], model.x) <= 84)

In [ ]:
lower_bound = 0
upper_bound = np.inf

In [ ]:
for j in col_indices:
  model.x[j].setlb(lower_bound)
  model.x[j].setub(upper_bound)

In [ ]:
model.pprint()

2 Set Declarations
    constraints_index : Size=1, Index=None, Ordered=Insertion
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :    2 : {1, 2}
    x_index : Size=1, Index=None, Ordered=Insertion
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :    5 : {0, 1, 2, 3, 4}

1 Var Declarations
    x : Size=5, Index=x_index
        Key : Lower : Value : Upper : Fixed : Stale : Domain
          0 :     0 :  None :  None : False :  True : NonNegativeIntegers
          1 :     0 :  None :  None : False :  True : NonNegativeIntegers
          2 :     0 :  None :  None : False :  True : NonNegativeIntegers
          3 :     0 :  None :  None : False :  True : NonNegativeIntegers
          4 :     0 :  None :  None : False :  True : NonNegativeIntegers

1 Objective Declarations
    objective : Size=1, Index=None, Active=True
        Key  : Active : Sense    : Expression
        None :   True : maximize : 11*x[0] + 13*x[1] + 9*x[2] + 15*x[3]

In [ ]:
!apt-get install -y -qq coinor-cbc

Selecting previously unselected package coinor-libcoinutils3v5.
(Reading database ... 155676 files and directories currently installed.)
Preparing to unpack .../0-coinor-libcoinutils3v5_2.10.14+repack1-1_amd64.deb ...
Unpacking coinor-libcoinutils3v5 (2.10.14+repack1-1) ...
Selecting previously unselected package coinor-libosi1v5.
Preparing to unpack .../1-coinor-libosi1v5_0.107.9+repack1-1_amd64.deb ...
Unpacking coinor-libosi1v5 (0.107.9+repack1-1) ...
Selecting previously unselected package coinor-libclp1.
Preparing to unpack .../2-coinor-libclp1_1.16.11+repack1-1_amd64.deb ...
Unpacking coinor-libclp1 (1.16.11+repack1-1) ...
Selecting previously unselected package coinor-libcgl1.
Preparing to unpack .../3-coinor-libcgl1_0.59.10+repack1-1_amd64.deb ...
Unpacking coinor-libcgl1 (0.59.10+repack1-1) ...
Selecting previously unselected package coinor-libcbc3.
Preparing to unpack .../4-coinor-libcbc3_2.9.9+repack1-1_amd64.deb ...
Unpacking coinor-libcbc3 (2.9.9+repack1-1) ...
Selecting p

In [ ]:
opt_cbc = SolverFactory('cbc')

In [ ]:
result = opt_cbc.solve(model)
print('Solver status:', result.solver.status)
print('Solver termination condition:',result.solver.termination_condition)

Solver status: ok
Solver termination condition: optimal


In [ ]:
# display solution
print('\nObjective = ', model.objective())

print('\nDecision Variables')
for i in col_indices:
  print('Num slots for Drone model ',data_csvfile['Drone model'][i], model.x[i].value)

print('\nConstraints')
model.constraints.display()


Objective =  132.0

Decision Variables
Num slots for Drone model  Shrey 12.0
Num slots for Drone model  Dharan 0.0
Num slots for Drone model  Satva 0.0
Num slots for Drone model  Adhya 0.0
Num slots for Drone model  Pravat 0.0

Constraints
constraints : Size=2
    Key : Lower : Body : Upper
      1 :  None : 78.0 :  78.0
      2 :  None : 69.0 :  84.0


Answer2:(b) $\\ $ The optimal solution is 132.0 = Maximum carrying capacity $ \\  $

 model  Shrey 12.0;
 
 model  Dharan 0.0;
 
 model Satva 0.0; 
 
 model Adhya 0.0;
 
 model Pravat 0.0; $ \\ $ 




Answerr2(c): We are to compare this obtained MILP solution with LP solution.For this we channge the domain of variables from Nonnegativeintegers to nonnegativereals.So:=

In [ ]:
model.x.domain = NonNegativeReals

In [ ]:
model.pprint()

2 Set Declarations
    constraints_index : Size=1, Index=None, Ordered=Insertion
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :    2 : {1, 2}
    x_index : Size=1, Index=None, Ordered=Insertion
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :    5 : {0, 1, 2, 3, 4}

1 Var Declarations
    x : Size=5, Index=x_index
        Key : Lower : Value : Upper : Fixed : Stale : Domain
          0 :     0 :  12.0 :  None : False : False : NonNegativeReals
          1 :     0 :   0.0 :  None : False : False : NonNegativeReals
          2 :     0 :   0.0 :  None : False : False : NonNegativeReals
          3 :     0 :   0.0 :  None : False : False : NonNegativeReals
          4 :     0 :   0.0 :  None : False : False : NonNegativeReals

1 Objective Declarations
    objective : Size=1, Index=None, Active=True
        Key  : Active : Sense    : Expression
        None :   True : maximize : 11*x[0] + 13*x[1] + 9*x[2] + 15*x[3] + 12*x[4]

1 C

In [ ]:
!apt-get install -y -qq coinor-cbc

In [ ]:
opt_cbc = SolverFactory('cbc')

In [ ]:
result = opt_cbc.solve(model)
print('Solver status:', result.solver.status)
print('Solver termination condition:',result.solver.termination_condition)

Solver status: ok
Solver termination condition: optimal


In [ ]:
# display solution
print('\nObjective = ', model.objective())

print('\nDecision Variables')
for i in col_indices:
  print('Num slots for Drone model ',data_csvfile['Drone model'][i], model.x[i].value)

print('\nConstraints')
model.constraints.display()


Objective =  132.0

Decision Variables
Num slots for Drone model  Shrey 12.0
Num slots for Drone model  Dharan 0.0
Num slots for Drone model  Satva 0.0
Num slots for Drone model  Adhya 0.0
Num slots for Drone model  Pravat 0.0

Constraints
constraints : Size=2
    Key : Lower : Body : Upper
      1 :  None : 78.0 :  78.0
      2 :  None : 69.0 :  84.0


Solution: Maximum carrying capacity = Objective =  132.0

 model  Shrey 12.0;
 
 model  Dharan 0.0;
 
 model Satva 0.0; 
 
 model Adhya 0.0;
 
 model Pravat 0.0;

Since now we again get same solution in LP as we get solution in MILP.It matches here but we can't say that it will be same in every condition like when we decrease value of limit of budget and increase value of parking space we find that optimal solution in LP and MILP are different.

Answer2(d): Now given that limit of budget is decreased s decreased to 76 crores of
rupees and the parking space is increased to 87 units. so now upper bounds of constraints are changed.

In [ ]:
model.constraints[1].deactivate()
model.constraints[2].deactivate()

In [ ]:
model.constraints.add(summation(data_csvfile['Cost (in crores of rupees)'], model.x) <= 76)
model.constraints.add(summation(data_csvfile['Parking space (sq.m.)'], model.x) <= 87)

In [ ]:
model.x.domain = NonNegativeReals

In [ ]:
model.pprint()

2 Set Declarations
    constraints_index : Size=1, Index=None, Ordered=Insertion
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :    4 : {1, 2, 3, 4}
    x_index : Size=1, Index=None, Ordered=Insertion
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :    5 : {0, 1, 2, 3, 4}

1 Var Declarations
    x : Size=5, Index=x_index
        Key : Lower : Value : Upper : Fixed : Stale : Domain
          0 :     0 :  12.0 :  None : False : False : NonNegativeReals
          1 :     0 :   0.0 :  None : False : False : NonNegativeReals
          2 :     0 :   0.0 :  None : False : False : NonNegativeReals
          3 :     0 :   0.0 :  None : False : False : NonNegativeReals
          4 :     0 :   0.0 :  None : False : False : NonNegativeReals

1 Objective Declarations
    objective : Size=1, Index=None, Active=True
        Key  : Active : Sense    : Expression
        None :   True : maximize : 11*x[0] + 13*x[1] + 9*x[2] + 15*x[3] + 12*x[4

In [ ]:
!apt-get install -y -qq coinor-cbc

In [ ]:
opt_cbc = SolverFactory('cbc')

In [ ]:
result = opt_cbc.solve(model)
print('Solver status:', result.solver.status)
print('Solver termination condition:',result.solver.termination_condition)

Solver status: ok
Solver termination condition: optimal


In [ ]:
# display solution
print('\nObjective = ', model.objective())

print('\nDecision Variables')
for i in col_indices:
  print('Num slots for Drone model ',data_csvfile['Drone model'][i], model.x[i].value)

print('\nConstraints')
model.constraints.display()


Objective =  128.615388

Decision Variables
Num slots for Drone model  Shrey 11.692308
Num slots for Drone model  Dharan 0.0
Num slots for Drone model  Satva 0.0
Num slots for Drone model  Adhya 0.0
Num slots for Drone model  Pravat 0.0

Constraints
constraints : Size=4
    Key : Lower : Body              : Upper
      3 :  None : 76.00000200000001 :  76.0
      4 :  None :         67.230771 :  87.0


Since as the limit of budget is decreased and the parking space is increased then value of objective function i.e maximum carrying capacity is decreased from 132 to 128.615388 kgs. and the value of variables also changed i.e number of drones of model shrey is decreased and number of other drones remains same (0).

Now we are to compare this solution to the MILP solution.So:=

In [ ]:
model.x.domain = NonNegativeIntegers

In [ ]:
model.pprint()

2 Set Declarations
    constraints_index : Size=1, Index=None, Ordered=Insertion
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :    4 : {1, 2, 3, 4}
    x_index : Size=1, Index=None, Ordered=Insertion
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :    5 : {0, 1, 2, 3, 4}

1 Var Declarations
    x : Size=5, Index=x_index
        Key : Lower : Value     : Upper : Fixed : Stale : Domain
          0 :     0 : 11.692308 :  None : False : False : NonNegativeIntegers
          1 :     0 :       0.0 :  None : False : False : NonNegativeIntegers
          2 :     0 :       0.0 :  None : False : False : NonNegativeIntegers
          3 :     0 :       0.0 :  None : False : False : NonNegativeIntegers
          4 :     0 :       0.0 :  None : False : False : NonNegativeIntegers

1 Objective Declarations
    objective : Size=1, Index=None, Active=True
        Key  : Active : Sense    : Expression
        None :   True : maximize : 11*x[0

In [ ]:
!apt-get install -y -qq coinor-cbc

In [ ]:
opt_cbc = SolverFactory('cbc')

In [ ]:
result = opt_cbc.solve(model)
print('Solver status:', result.solver.status)
print('Solver termination condition:',result.solver.termination_condition)

Solver status: ok
Solver termination condition: optimal


In [ ]:
# display solution
print('\nObjective = ', model.objective())

print('\nDecision Variables')
for i in col_indices:
  print('Num slots for Drone model ',data_csvfile['Drone model'][i], model.x[i].value)

print('\nConstraints')
model.constraints.display()


Objective =  127.0

Decision Variables
Num slots for Drone model  Shrey 8.0
Num slots for Drone model  Dharan 3.0
Num slots for Drone model  Satva 0.0
Num slots for Drone model  Adhya 0.0
Num slots for Drone model  Pravat 0.0

Constraints
constraints : Size=4
    Key : Lower : Body : Upper
      3 :  None : 76.0 :  76.0
      4 :  None : 59.5 :  87.0


So we can see that as we put a restriction on domain of variables of integers then we see that maximum carrying capacity decreases to 127 from  128.615388.And also number of drones of model shrey also decreases to 8 from  11.692308 and number of drones of model Dharan increases to 3 from 0.and other ramains same.

Answer2(e): Now we are to decrease limit of budget from 76 crores to 74, 72, 70 and 68 crores of rupees and increase parking sapce e to 88, 89, 90 and 91 units respectively.

Case1: limit of budget = 74; parking space = 88



In [ ]:
model.constraints[3].deactivate()
model.constraints[4].deactivate()

In [ ]:
model.constraints.add(summation(data_csvfile['Cost (in crores of rupees)'], model.x) <= 74)
model.constraints.add(summation(data_csvfile['Parking space (sq.m.)'], model.x) <= 88)

In [ ]:
model.x.domain = NonNegativeReals

In [ ]:
model.pprint()

2 Set Declarations
    constraints_index : Size=1, Index=None, Ordered=Insertion
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :    6 : {1, 2, 3, 4, 5, 6}
    x_index : Size=1, Index=None, Ordered=Insertion
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :    5 : {0, 1, 2, 3, 4}

1 Var Declarations
    x : Size=5, Index=x_index
        Key : Lower : Value : Upper : Fixed : Stale : Domain
          0 :     0 :   8.0 :  None : False : False : NonNegativeReals
          1 :     0 :   3.0 :  None : False : False : NonNegativeReals
          2 :     0 :   0.0 :  None : False : False : NonNegativeReals
          3 :     0 :   0.0 :  None : False : False : NonNegativeReals
          4 :     0 :   0.0 :  None : False : False : NonNegativeReals

1 Objective Declarations
    objective : Size=1, Index=None, Active=True
        Key  : Active : Sense    : Expression
        None :   True : maximize : 11*x[0] + 13*x[1] + 9*x[2] + 15*x[3] + 

In [ ]:
!apt-get install -y -qq coinor-cbc

In [ ]:
opt_cbc = SolverFactory('cbc')


In [ ]:
result = opt_cbc.solve(model)
print('Solver status:', result.solver.status)
print('Solver termination condition:',result.solver.termination_condition)

Solver status: ok
Solver termination condition: optimal


In [ ]:
# display solution
print('\nObjective = ', model.objective())

print('\nDecision Variables')
for i in col_indices:
  print('Num slots for Drone model ',data_csvfile['Drone model'][i], model.x[i].value)

print('\nConstraints')
model.constraints.display()


Objective =  125.230765

Decision Variables
Num slots for Drone model  Shrey 11.384615
Num slots for Drone model  Dharan 0.0
Num slots for Drone model  Satva 0.0
Num slots for Drone model  Adhya 0.0
Num slots for Drone model  Pravat 0.0

Constraints
constraints : Size=6
    Key : Lower : Body        : Upper
      5 :  None :  73.9999975 :  74.0
      6 :  None : 65.46153625 :  88.0


As LP we get solution 125.230765.

In [ ]:
model.x.domain = NonNegativeIntegers

In [ ]:
model.pprint()

2 Set Declarations
    constraints_index : Size=1, Index=None, Ordered=Insertion
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :    6 : {1, 2, 3, 4, 5, 6}
    x_index : Size=1, Index=None, Ordered=Insertion
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :    5 : {0, 1, 2, 3, 4}

1 Var Declarations
    x : Size=5, Index=x_index
        Key : Lower : Value     : Upper : Fixed : Stale : Domain
          0 :     0 : 11.384615 :  None : False : False : NonNegativeIntegers
          1 :     0 :       0.0 :  None : False : False : NonNegativeIntegers
          2 :     0 :       0.0 :  None : False : False : NonNegativeIntegers
          3 :     0 :       0.0 :  None : False : False : NonNegativeIntegers
          4 :     0 :       0.0 :  None : False : False : NonNegativeIntegers

1 Objective Declarations
    objective : Size=1, Index=None, Active=True
        Key  : Active : Sense    : Expression
        None :   True : maximize : 

In [ ]:
!apt-get install -y -qq coinor-cbc

In [ ]:
opt_cbc = SolverFactory('cbc')

In [ ]:
result = opt_cbc.solve(model)
print('Solver status:', result.solver.status)
print('Solver termination condition:',result.solver.termination_condition)

Solver status: ok
Solver termination condition: optimal


In [ ]:
# display solution
print('\nObjective = ', model.objective())

print('\nDecision Variables')
for i in col_indices:
  print('Num slots for Drone model ',data_csvfile['Drone model'][i], model.x[i].value)

print('\nConstraints')
model.constraints.display()


Objective =  123.0

Decision Variables
Num slots for Drone model  Shrey 10.0
Num slots for Drone model  Dharan 1.0
Num slots for Drone model  Satva 0.0
Num slots for Drone model  Adhya 0.0
Num slots for Drone model  Pravat 0.0

Constraints
constraints : Size=6
    Key : Lower : Body : Upper
      5 :  None : 73.0 :  74.0
      6 :  None : 62.0 :  88.0


As MILP we get maximum carrying capacity 123kgs

Case 2: limit of budget = 72; parking space = 89

In [ ]:
model.constraints[5].deactivate()
model.constraints[6].deactivate()

In [ ]:
model.constraints.add(summation(data_csvfile['Cost (in crores of rupees)'], model.x) <= 72)
model.constraints.add(summation(data_csvfile['Parking space (sq.m.)'], model.x) <= 89)

In [ ]:
model.x.domain = NonNegativeReals

In [ ]:
model.pprint()

2 Set Declarations
    constraints_index : Size=1, Index=None, Ordered=Insertion
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :    8 : {1, 2, 3, 4, 5, 6, 7, 8}
    x_index : Size=1, Index=None, Ordered=Insertion
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :    5 : {0, 1, 2, 3, 4}

1 Var Declarations
    x : Size=5, Index=x_index
        Key : Lower : Value : Upper : Fixed : Stale : Domain
          0 :     0 :  10.0 :  None : False : False : NonNegativeReals
          1 :     0 :   1.0 :  None : False : False : NonNegativeReals
          2 :     0 :   0.0 :  None : False : False : NonNegativeReals
          3 :     0 :   0.0 :  None : False : False : NonNegativeReals
          4 :     0 :   0.0 :  None : False : False : NonNegativeReals

1 Objective Declarations
    objective : Size=1, Index=None, Active=True
        Key  : Active : Sense    : Expression
        None :   True : maximize : 11*x[0] + 13*x[1] + 9*x[2] + 15*x

In [ ]:
!apt-get install -y -qq coinor-cbc

In [ ]:
opt_cbc = SolverFactory('cbc')

In [ ]:
result = opt_cbc.solve(model)
print('Solver status:', result.solver.status)
print('Solver termination condition:',result.solver.termination_condition)

Solver status: ok
Solver termination condition: optimal


In [ ]:
# display solution
print('\nObjective = ', model.objective())

print('\nDecision Variables')
for i in col_indices:
  print('Num slots for Drone model ',data_csvfile['Drone model'][i], model.x[i].value)

print('\nConstraints')
model.constraints.display()


Objective =  121.84615300000002

Decision Variables
Num slots for Drone model  Shrey 11.076923
Num slots for Drone model  Dharan 0.0
Num slots for Drone model  Satva 0.0
Num slots for Drone model  Adhya 0.0
Num slots for Drone model  Pravat 0.0

Constraints
constraints : Size=8
    Key : Lower : Body               : Upper
      7 :  None :         71.9999995 :  72.0
      8 :  None : 63.692307250000006 :  89.0


We get maximum carrying capacity = 121.846153

In [ ]:
model.x.domain = NonNegativeIntegers

In [ ]:
model.pprint()

2 Set Declarations
    constraints_index : Size=1, Index=None, Ordered=Insertion
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :    8 : {1, 2, 3, 4, 5, 6, 7, 8}
    x_index : Size=1, Index=None, Ordered=Insertion
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :    5 : {0, 1, 2, 3, 4}

1 Var Declarations
    x : Size=5, Index=x_index
        Key : Lower : Value     : Upper : Fixed : Stale : Domain
          0 :     0 : 11.076923 :  None : False : False : NonNegativeIntegers
          1 :     0 :       0.0 :  None : False : False : NonNegativeIntegers
          2 :     0 :       0.0 :  None : False : False : NonNegativeIntegers
          3 :     0 :       0.0 :  None : False : False : NonNegativeIntegers
          4 :     0 :       0.0 :  None : False : False : NonNegativeIntegers

1 Objective Declarations
    objective : Size=1, Index=None, Active=True
        Key  : Active : Sense    : Expression
        None :   True : maxim

In [ ]:
!apt-get install -y -qq coinor-cbc

In [ ]:
opt_cbc = SolverFactory('cbc')

In [ ]:
result = opt_cbc.solve(model)
print('Solver status:', result.solver.status)
print('Solver termination condition:',result.solver.termination_condition)

Solver status: ok
Solver termination condition: optimal


In [ ]:
# display solution
print('\nObjective = ', model.objective())

print('\nDecision Variables')
for i in col_indices:
  print('Num slots for Drone model ',data_csvfile['Drone model'][i], model.x[i].value)

print('\nConstraints')
model.constraints.display()


Objective =  121.0

Decision Variables
Num slots for Drone model  Shrey 11.0
Num slots for Drone model  Dharan 0.0
Num slots for Drone model  Satva 0.0
Num slots for Drone model  Adhya 0.0
Num slots for Drone model  Pravat 0.0

Constraints
constraints : Size=8
    Key : Lower : Body  : Upper
      7 :  None :  71.5 :  72.0
      8 :  None : 63.25 :  89.0


So as MILP we get  that maximum carrying capacity = 121

Case3:  limit of budget = 70; parking space = 90

In [ ]:
model.constraints[7].deactivate()
model.constraints[8].deactivate()

In [ ]:
model.constraints.add(summation(data_csvfile['Cost (in crores of rupees)'], model.x) <= 70)
model.constraints.add(summation(data_csvfile['Parking space (sq.m.)'], model.x) <= 90)

In [ ]:
model.x.domain = NonNegativeReals

In [ ]:
model.pprint()

2 Set Declarations
    constraints_index : Size=1, Index=None, Ordered=Insertion
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :   10 : {1, 2, 3, 4, 5, 6, 7, 8, 9, 10}
    x_index : Size=1, Index=None, Ordered=Insertion
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :    5 : {0, 1, 2, 3, 4}

1 Var Declarations
    x : Size=5, Index=x_index
        Key : Lower : Value : Upper : Fixed : Stale : Domain
          0 :     0 :  11.0 :  None : False : False : NonNegativeReals
          1 :     0 :   0.0 :  None : False : False : NonNegativeReals
          2 :     0 :   0.0 :  None : False : False : NonNegativeReals
          3 :     0 :   0.0 :  None : False : False : NonNegativeReals
          4 :     0 :   0.0 :  None : False : False : NonNegativeReals

1 Objective Declarations
    objective : Size=1, Index=None, Active=True
        Key  : Active : Sense    : Expression
        None :   True : maximize : 11*x[0] + 13*x[1] + 9*x[2]

In [ ]:
!apt-get install -y -qq coinor-cbc

In [ ]:
opt_cbc = SolverFactory('cbc')

In [ ]:
result = opt_cbc.solve(model)
print('Solver status:', result.solver.status)
print('Solver termination condition:',result.solver.termination_condition)

Solver status: ok
Solver termination condition: optimal


In [ ]:
# display solution
print('\nObjective = ', model.objective())

print('\nDecision Variables')
for i in col_indices:
  print('Num slots for Drone model ',data_csvfile['Drone model'][i], model.x[i].value)

print('\nConstraints')
model.constraints.display()


Objective =  118.461541

Decision Variables
Num slots for Drone model  Shrey 10.769231
Num slots for Drone model  Dharan 0.0
Num slots for Drone model  Satva 0.0
Num slots for Drone model  Adhya 0.0
Num slots for Drone model  Pravat 0.0

Constraints
constraints : Size=10
    Key : Lower : Body               : Upper
      9 :  None :         70.0000015 :  70.0
     10 :  None : 61.923078249999996 :  90.0


As LP we get maximum carrying capacity = 118.461541

In [ ]:
model.x.domain = NonNegativeIntegers

In [ ]:
model.pprint()

2 Set Declarations
    constraints_index : Size=1, Index=None, Ordered=Insertion
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :   10 : {1, 2, 3, 4, 5, 6, 7, 8, 9, 10}
    x_index : Size=1, Index=None, Ordered=Insertion
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :    5 : {0, 1, 2, 3, 4}

1 Var Declarations
    x : Size=5, Index=x_index
        Key : Lower : Value     : Upper : Fixed : Stale : Domain
          0 :     0 : 10.769231 :  None : False : False : NonNegativeIntegers
          1 :     0 :       0.0 :  None : False : False : NonNegativeIntegers
          2 :     0 :       0.0 :  None : False : False : NonNegativeIntegers
          3 :     0 :       0.0 :  None : False : False : NonNegativeIntegers
          4 :     0 :       0.0 :  None : False : False : NonNegativeIntegers

1 Objective Declarations
    objective : Size=1, Index=None, Active=True
        Key  : Active : Sense    : Expression
        None :   True 

In [ ]:
!apt-get install -y -qq coinor-cbc

In [ ]:
opt_cbc = SolverFactory('cbc')

In [ ]:
result = opt_cbc.solve(model)
print('Solver status:', result.solver.status)
print('Solver termination condition:',result.solver.termination_condition)

Solver status: ok
Solver termination condition: optimal


In [ ]:
# display solution
print('\nObjective = ', model.objective())

print('\nDecision Variables')
for i in col_indices:
  print('Num slots for Drone model ',data_csvfile['Drone model'][i], model.x[i].value)

print('\nConstraints')
model.constraints.display()


Objective =  116.0

Decision Variables
Num slots for Drone model  Shrey 7.0
Num slots for Drone model  Dharan 3.0
Num slots for Drone model  Satva 0.0
Num slots for Drone model  Adhya 0.0
Num slots for Drone model  Pravat 0.0

Constraints
constraints : Size=10
    Key : Lower : Body  : Upper
      9 :  None :  69.5 :  70.0
     10 :  None : 53.75 :  90.0


As MILP we get maximum carrying capacity = 116 kgs

 Case4:limit of budget = 68; parking space = 91

In [ ]:
model.constraints[9].deactivate()
model.constraints[10].deactivate()

In [ ]:
model.constraints.add(summation(data_csvfile['Cost (in crores of rupees)'], model.x) <= 68)
model.constraints.add(summation(data_csvfile['Parking space (sq.m.)'], model.x) <= 91)

In [ ]:
model.x.domain = NonNegativeReals

In [ ]:
model.pprint()

2 Set Declarations
    constraints_index : Size=1, Index=None, Ordered=Insertion
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :   12 : {1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12}
    x_index : Size=1, Index=None, Ordered=Insertion
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :    5 : {0, 1, 2, 3, 4}

1 Var Declarations
    x : Size=5, Index=x_index
        Key : Lower : Value : Upper : Fixed : Stale : Domain
          0 :     0 :   7.0 :  None : False : False : NonNegativeReals
          1 :     0 :   3.0 :  None : False : False : NonNegativeReals
          2 :     0 :   0.0 :  None : False : False : NonNegativeReals
          3 :     0 :   0.0 :  None : False : False : NonNegativeReals
          4 :     0 :   0.0 :  None : False : False : NonNegativeReals

1 Objective Declarations
    objective : Size=1, Index=None, Active=True
        Key  : Active : Sense    : Expression
        None :   True : maximize : 11*x[0] + 13*x[1] 

In [ ]:
!apt-get install -y -qq coinor-cbc

In [ ]:
opt_cbc = SolverFactory('cbc')

In [ ]:
result = opt_cbc.solve(model)
print('Solver status:', result.solver.status)
print('Solver termination condition:',result.solver.termination_condition)

Solver status: ok
Solver termination condition: optimal


In [ ]:
# display solution
print('\nObjective = ', model.objective())

print('\nDecision Variables')
for i in col_indices:
  print('Num slots for Drone model ',data_csvfile['Drone model'][i], model.x[i].value)

print('\nConstraints')
model.constraints.display()


Objective =  115.07691799999999

Decision Variables
Num slots for Drone model  Shrey 10.461538
Num slots for Drone model  Dharan 0.0
Num slots for Drone model  Satva 0.0
Num slots for Drone model  Adhya 0.0
Num slots for Drone model  Pravat 0.0

Constraints
constraints : Size=12
    Key : Lower : Body               : Upper
     11 :  None :          67.999997 :  68.0
     12 :  None : 60.153843499999994 :  91.0


As LP we get maximum carrying capacity = 115.07691799999

In [ ]:
model.x.domain = NonNegativeIntegers

In [ ]:
model.pprint()

2 Set Declarations
    constraints_index : Size=1, Index=None, Ordered=Insertion
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :   12 : {1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12}
    x_index : Size=1, Index=None, Ordered=Insertion
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :    5 : {0, 1, 2, 3, 4}

1 Var Declarations
    x : Size=5, Index=x_index
        Key : Lower : Value     : Upper : Fixed : Stale : Domain
          0 :     0 : 10.461538 :  None : False : False : NonNegativeIntegers
          1 :     0 :       0.0 :  None : False : False : NonNegativeIntegers
          2 :     0 :       0.0 :  None : False : False : NonNegativeIntegers
          3 :     0 :       0.0 :  None : False : False : NonNegativeIntegers
          4 :     0 :       0.0 :  None : False : False : NonNegativeIntegers

1 Objective Declarations
    objective : Size=1, Index=None, Active=True
        Key  : Active : Sense    : Expression
        None :

In [ ]:
!apt-get install -y -qq coinor-cbc

In [ ]:
opt_cbc = SolverFactory('cbc')

In [ ]:
result = opt_cbc.solve(model)
print('Solver status:', result.solver.status)
print('Solver termination condition:',result.solver.termination_condition)

Solver status: ok
Solver termination condition: optimal


In [ ]:
# display solution
print('\nObjective = ', model.objective())

print('\nDecision Variables')
for i in col_indices:
  print('Num slots for Drone model ',data_csvfile['Drone model'][i], model.x[i].value)

print('\nConstraints')
model.constraints.display()


Objective =  114.0

Decision Variables
Num slots for Drone model  Shrey 8.0
Num slots for Drone model  Dharan 2.0
Num slots for Drone model  Satva 0.0
Num slots for Drone model  Adhya 0.0
Num slots for Drone model  Pravat 0.0

Constraints
constraints : Size=12
    Key : Lower : Body : Upper
     11 :  None : 68.0 :  68.0
     12 :  None : 55.0 :  91.0


As MILP we get maximum carrying capacity = 114 kgs

(As MILP):maximum carrying capacity  = 123kgs,121kgs,116kgs,114kgs when limit of budget = 74,72,70,68 and parking space = 88,89,90,91 respectively.

(As LP):maximum carrying capacity  = 125.230765
kgs,121.84615300000002kgs,118.461541kgs,115.07691799999kgs when limit of budget = 74,72,70,68 and parking space = 88,89,90,91 respectively


By above obtained data we can conclude that as limit of budget decreases and parking space increases then carrying capacity decreases.
 


My observation on obtained solution is that if limit of budget is decreased and parking space is increased then only limit of budget puts effect in decreasing optimal solution since all the times constraint related to parking space is not active but constraint related to limit of budget is always active constraint.that is reason behind getting such kind of solution.